src/hlsBinaries.ts

```diff
@@ -247,45 +256,63 @@ export async function findHaskellLanguageServer(
     return findHLSinPATH(context, logger, folder);
   } else {
     // we manage HLS, make sure ghcup is installed/available
-    await getGHCup(context, logger);
-
-    // get a preliminary hls wrapper for finding project GHC version,
-    // later we may install a different HLS that supports the given GHC
-    let wrapper = await getLatestWrapperFromGHCup(context, logger).then((e) =>
-      !e
-        ? callGHCup(context, logger, ['install', 'hls'], 'Installing latest HLS', true).then(() =>
-            callGHCup(
-              context,
-              logger,
-              ['whereis', 'hls'],
-              undefined,
-              false,
-              (err, stdout, _stderr, resolve, reject) => {
-                err ? reject("Couldn't find latest HLS") : resolve(stdout?.trim());
-              }
-            )
-          )
-        : e
+    await upgradeGHCup(context, logger);
+
+    // get a preliminary toolchain for finding the correct project GHC version
+    // (we need HLS and cabal/stack and ghc as fallback),
+    // later we may install a different toolchain that's more project-specific
+    const latestHLS = await getLatestToolFromGHCup(context, logger, 'hls');
+    const latestCabal = (workspace.getConfiguration('haskell').get('installCabal') as boolean)
+      ? await getLatestToolFromGHCup(context, logger, 'cabal')
+      : null;
+    const latestStack = (workspace.getConfiguration('haskell').get('installStack') as boolean)
+      ? await getLatestToolFromGHCup(context, logger, 'stack')
+      : null;
+    const recGHC =
+      !(await executableExists('ghc')) && (workspace.getConfiguration('haskell').get('installGHC') as boolean)
+        ? await getLatestAvailableToolFromGHCup(context, logger, 'ghc', 'recommended')
+        : null;
```

Comment:
Slightly unrelated, we should have a notion of a ghc is "recommended" and "hls-powered".

This parses for me as if "ghc does not exist or installGhc is true", then install the recommended ghc version. This behaviour is noticeably different to stack and cabal, because if those do not exist, then we don't override the config to automatically install stack and cabal, but rather we crash. I believe, it should be the same for `ghc`.

Reply:
Ok, I mentally misparsed the parenthesis. It is, if ghc is not on the PATH and installGHC is enabled, then install the recommended one. This is fine again